Add Yiming's GSoC blog (#5402)
jenkins-infra/jenkins.io: content/blog/2022/08/2022-08-24-jenkinsfile-runner-as-github-actions.adoc

---
layout: post
title: "A near Feature-complete version of Jenkinsfile Runner Actions as GitHub Actions"
tags:
- gsoc
- gsoc2022
- jenkinsfile-runner
- github-actions
author: YimingGong
---

== Abstract

Jenkinsfile Runner Action for GitHub Actions aims at providing customized environments and useful GitHub Actions for the users.
Therefore, users can execute the Jenkins pipeline in the GitHub Actions runners.

> It wants to run Jenkins pipeline in a Function-as-a-Service context.

Reviewer: ```suggestion
It enables running Jenkins pipelines in a Function-as-a-Service context.
```

Probably also link to https://en.wikipedia.org/wiki/Function_as_a_Service so that people not familiar with FaaS have a better understanding.

Users can customize their environment by setting up base images, configuring the Jenkins Configuration-as-Code plugin,
extending vanilla images or providing initial hooks for the temporary Jenkins controller.

> It also provides some examples about how to set up secrets, Jenkins plugins and integrate with other GitHub Actions.

Reviewer: ```suggestion
Several examples about how to set up secrets, Jenkins plugins and integrate with other GitHub Actions can be seen below.
```

* `jfr-runtime-action` - It runs the Jenkins pipeline in the host machine. `jfr-setup-setup` needs to be run before.

=== Container type actions

These actions run the pipeline in the customized container.

* `jfr-container-action` - It runs the Jenkins pipeline in the declared container.
As the prebuilt container is declared in advance, it can integrate with other GitHub Actions theoretically.
It's based on the idea of link:https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container[Running jobs in a container].

* `jfr-static-image-action` - It runs the Jenkins pipeline in the runtime built container.
It doesn't need to declare the container in advance so users cannot use other GitHub Actions except using `actions/checkout` to set up the workspace.
But the users can override the base image by providing the docker image URL.
For instance, if users want to combine the Jenkins environment with nodejs environment, they can use `node:18.3.0` as input.
It's based on the idea of link:https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action#introduction/[GitHub Docker container action].

== Phase 1

> When I started to write codes based on my previous proposal, I found my implementation is different from the expectation of the maintainer and the mentors.

Reviewer: ```suggestion
When I started to write code based on my previous proposal, I found my implementation is different from the expectation of the maintainer and the mentors.
```

          pluginstxt: plugins.txt
          jcasc: jcasc.yml 
----

=== How to use jfr-runtime-action with other supporting runtime type actions

`jfr-runtime-action` will run the pipeline in the host machine directly.
The users need to use `jfr-setup-action` in advance.
If the users want to install extra plugins, they can use `jfr-plugin-installation-action`.
The advantage of `jfr-runtime-action` is that it can run in all kinds of runners provided by GitHub Actions.
[source,yaml]
----
name: CI
on: [push]
jobs:
  jfr-runtime-action-pipeline:
    strategy:
      matrix:
        os: [ ubuntu-latest, macOS-latest, windows-latest ]
    runs-on: ${{matrix.os}}
    name: jfr-runtime-action-pipeline
    steps:
      - uses: actions/checkout@v2
      - name : Setup Jenkins
        uses:
          jenkinsci/jfr-runtime-action@master
      - name: Jenkins plugins download
        uses:
          jenkinsci/jfr-plugin-installation-action@master
        with:
          pluginstxt: plugins.txt
      - name: Run Jenkins pipeline
        uses:
          jenkinsci/jfr-runtime-action@master
        with:
          command: run
          jenkinsfile: Jenkinsfile
----

== Reflection

> The most valuable lesson I learned in GSoC is not about how to write codes.

Reviewer 2: ```suggestion
The most valuable lesson I learned in GSoC is not about how to write code.
```
